fontforgeexe/cvpalettes.c

```diff
@@ -819,100 +817,281 @@ static void CVPolyStar(CharView *cv) {
+	{ &GIcon_flip_selected,        &GIcon_skew_selected },
+	{ &GIcon_3drotate_selected,    &GIcon_perspective_selected },
+	{ &GIcon_rect_selected,        &GIcon_poly_selected},
+	{ 0, 0 },
+	{ &GIcon_elipse_selected,      &GIcon_star_selected},
+	{ 0, 0 },
+	{ 0, 0 }
+    }};
+static GImage *normsmalls[] = { &GIcon_smallpointer,  &GIcon_smallmag,
+				&GIcon_smallpencil,   &GIcon_smallhand,
+				&GIcon_smallknife,    &GIcon_smallruler,
+				&GIcon_smallpen,      NULL,
+				&GIcon_smallcurve,    &GIcon_smallhvcurve,
+				&GIcon_smallcorner,   &GIcon_smalltangent,
+				&GIcon_smallscale,    &GIcon_smallrotate,
+				&GIcon_smallflip,     &GIcon_smallskew,
+				&GIcon_small3drotate, &GIcon_smallperspective,
+				&GIcon_smallrect,     &GIcon_smallpoly,
+				&GIcon_smallelipse,   &GIcon_smallstar };
+static GImage *spirosmalls[] = { &GIcon_smallpointer, &GIcon_smallmag,
+				 &GIcon_smallpencil,  &GIcon_smallhand,
+				 &GIcon_smallknife,   &GIcon_smallruler,
+				 &GIcon_smallspiroright,  NULL,
+				 &GIcon_smallspirocurve,  &GIcon_smallspirog2curve,
+				 &GIcon_smallspirocorner, &GIcon_smallspiroleft,
+				 &GIcon_smallscale,       &GIcon_smallrotate,
+				 &GIcon_smallflip,        &GIcon_smallskew,
+				 &GIcon_small3drotate,    &GIcon_smallperspective,
+				 &GIcon_smallrect,        &GIcon_smallpoly,
+				 &GIcon_smallelipse,      &GIcon_smallstar };
+
+static int getSmallIconsHeight()
+{
+    return GIcon_smallpointer.u.image->height;
+}
+
+static int getToolbarWidth( CharView *cv ) 
+{
+    int cache = 0;
+    if( !cache ) {
+	GImage* (*buttons)[14][2] = (CVInSpiro(cv) ? spirobuttons : normbuttons);
+	int i = 0;
+	
+	for ( i=0; buttons[0][i][0]; ++i ) {
+	    cache = MAX( cache, buttons[0][i][0]->u.image->width + buttons[0][i][1]->u.image->width );
+	}
+    }
+    return cache;
+}
+
+static int getToolbarHeight( CharView *cv ) 
+{
+    int cache = 0;
+    if( !cache ) {
+	GImage* (*buttons)[14][2] = (CVInSpiro(cv) ? spirobuttons : normbuttons);
+	int i = 0;
+	
+	for ( i=0; buttons[0][i][0]; ++i ) {
+	    cache += MAX( buttons[0][i][0]->u.image->height, buttons[0][i][1]->u.image->height );
+	}
+    }
+    cache += getSmallIconsHeight() * 4;
+    cache += 6;
+    return cache;
+}
+
+
+typedef struct _IJ 
+{
+    int i;
+    int j;
+} IJ;
+
+typedef void (*visitButtonsVisitor) ( CharView *cv,             // CharView passed to visitButtons()
+				      GImage* gimage, int mi,   // button we are visiting and adjusted 'i'
+				      int i, int j,             // i and j position of button (j is across 0,1) (i is down 0...11)
+				      int iconx, int icony,     // pixel where this button starts
+				      int selected,             // is this button the active tool
+				      void* udata );            // user data
+
+
+/**
+ * Visit every button in the toolbar calling the function 'v' with a
+ * collection of interesting state data.
+ */
+static void visitButtons( CharView* cv, visitButtonsVisitor v, void* udata ) 
+{
```

Comment:
It would be very helpful to have comments here explaining what happens. The logic is not particularly obvious.